sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala

```diff
@@ -106,11 +162,11 @@ case class Mode(
     val collationAwareBuffer = child.dataType match {
```

Comment:
```suggestion
    val collationAwareBuffer = getCollationAwareBuffer(buffer)
```

let's separate this out

Reply:
and when doing that in a separate function, we should try to follow a logic similar to `RewriteCollationJoin`

`getCollationAwareBuffer` can do something like (rough sketch):
```
    dataType match {
      // Short-circuit if there is no collation.
      case _ if UnsafeRowUtils.isBinaryStable(child.dataType) => ...
      case st: StringType =>
        getCollationAwareBufferForStringType(...)
      case at: ArrayType =>
        getCollationAwareBufferForArrayType(...)
      case  st : StructType =>
        getCollationAwareBufferForStructType(...)
      // Write a comment to point out that don't support MapType (and possibly some other types)...
      case _ => buffer
    }
```

and then you can have:
```
def getCollationAwareBufferForStringType ...
def getCollationAwareBufferForArrayType ...
def getCollationAwareBufferForStructType ...
```
that could all rely on `getCollationAwareBuffer` instead of having to call each other

for example, imagine if we add support MapType, then:
1. recursivelyGetBufferForStructType would need a case for ArrayType & MapType
2. recursivelyGetBufferForArrayType would need a case for StructType & MapType
3. recursivelyGetBufferForMapType would need a case for ArrayType & StructType
(you see where this is going)